docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+
+If the `None` value is used (rather than just being pattern matched on and ignored) then the compiler must be able to materialise the address (or capability) to the singleton.
+This is no worse than if we stored the address inline; however, because in that case the match would require a comparison against the address and so the address would need to be loaded before the pattern match, rather than conditionally in one path.
+
+Objects
+-------
+
+Verona objects contain a header with metadata describing the object, followed by user-defined fields.
+
+### Object headers
+
+Object headers are two pointer-size words.
+These are described in detail in [src/rt/object/object.h](../src/rt/object/object.h).
+
+The first word contains the region metadata.
+This is a discriminated union where the low three bits describe how the remainder is used (for example, as a reference count for one immutable object in an SCC, a pointer to that object for other immutable objects in an SCC, and a pointer to the region metadata for sentinel objects).
+
+The second is a pointer to the descriptor for the object.
+The low bits of this are used by the runtime for memory management state and so must be zeroed before accesses.
+
+The object descriptor layout is still likely to change, the current layout is:
+
+64-bit integer size.
+Pointer to a trace function, used in traced regions to walk reachable objects.
+Pointer to a finalise function, called when an object is dead but before other objects in the same region are deleted (though they may have already been finalised.
+Pointer to a notify function, invoked when an object is notified of a noticeboard event.
+Pointer to a destructor, which is called after some objects in a region have been deallocated and may not follow any `mut` pointers.
+The remainder of the descriptor is the vtable.
+
+Methods in the vtable use *selector colouring*.
+Each selector (method name and types) is assigned a separate vtable index.
+Pairs of selectors that do not appear in the same interface are then combined to use the same index.
+Selector colouring is not essential but can dramatically reduce the size of vtables.
+
+*Note:* If we can distinguish between sentinel objects and others at object-creation time, the region metadata field for non-sentinel objects could be a 32-bit value that is either an integer or an index into a table, rather than a pointer.
+This would be a noticeable saving on CHERI systems and a small saving on 64-bit systems, at the cost of an extra indirection for some memory-management operations.
+We could also use a 32-bit index into a descriptor table, as the Fuchsia C++ ABI does: Programs that have more than 2^32 distinct types arise only as a result of poor life choices.
+
+*Note:* The above choices are visible to the runtime but are part of the internal ABI and so it would be possible to have size and performance optimised variants available on a per-program basis.
```

Comment:
In theory, yes, though I don't know if we can automatically differentiate between objects that are touched only in cold code paths and ones that aren't.  We might be able to do this on a per-type basis with explicit programmer opt-in, though we might end up needing to steal one bit in the header to differentiate between the two.

Reply:
Ah, yes! Otherwise we'd create two implementations of the same conversion and that would be against the core design goal of being able to recompute, cache, etc.

The hot/cold part could be done using profile information or user pragmas, but also, this is subject for another discussion.